src/python/pants/engine/target.py

```diff
@@ -1612,12 +1616,20 @@ def can_generate(
     def path_globs(self, files_not_found_behavior: FilesNotFoundBehavior) -> PathGlobs:
         if not self.globs:
             return PathGlobs([])
-        error_behavior = files_not_found_behavior.to_glob_match_error_behavior()
+        # Match any if we use default globs, else match all.
         conjunction = (
             GlobExpansionConjunction.all_match
-            if not self.default or (set(self.globs) != set(self.default))
+            if not self.default or (set(self.value) != set(self.default))  # type: ignore[arg-type]
             else GlobExpansionConjunction.any_match
         )
+        # Use fields default error behavior if defined, if we use default globs else the provided
+        # error behavior.
+        error_behavior = (
+            files_not_found_behavior.to_glob_match_error_behavior()
+            if conjunction == GlobExpansionConjunction.all_match
+            or self.default_glob_match_error_behavior is None
+            else self.default_glob_match_error_behavior
+        )
```

Comment:
Just realized that it is perhaps better to simply do a single if-else assigning to both these variables (conjunction and error_behavior). Can amend tomorrow.

Reply:
Nope, that didn't improve the readability or look of this section. Leaving as-is is better.